internal/upload/ftp.go

```diff
@@ -203,57 +112,17 @@ func (agent *FTPTransferAgent) Hostname() string {
 }
 
 func (agent *FTPTransferAgent) Delete(path string) error {
-	agent.mu.Lock()
-	defer agent.mu.Unlock()
-
-	if path == "" || strings.HasSuffix(path, "/") {
-		return fmt.Errorf("FTPTransferAgent: invalid path %v", path)
-	}
-
-	conn, err := agent.connection()
-	if err != nil {
-		return err
-	}
-
-	err = conn.Delete(path)
-	if err != nil && !strings.Contains(err.Error(), "no such file or directory") {
-		return err
-	}
-	return nil
+	return agent.client.Delete(path)
 }
 
 // uploadFile saves the content of File at the given filename in the OutboundPath directory
 //
 // The File's contents will always be closed
 func (agent *FTPTransferAgent) UploadFile(f File) error {
-	defer f.Close()
-
-	agent.mu.Lock()
-	defer agent.mu.Unlock()
-
-	conn, err := agent.connection()
-	if err != nil {
-		return err
-	}
-
-	// move into inbound directory and set a trigger to undo and set a defer to move back
-	wd, err := conn.CurrentDir()
-	if err != nil {
-		return err
-	}
-	if err := conn.ChangeDir(agent.cfg.Paths.Outbound); err != nil {
-		return err
+	if agent == nil || agent.cfg.FTP == nil {
```

Comment:
What is the purpose of the nil check on the receiver (agent)? Without this nil check I assume that the result would be a panic. How is returning an error different from a panic? I have done some googling, but can't find any very good answers to the questions of whether a nil check is needed when the receiver is a pointer. The best answer I have found says it is better to let it panic since calling a method on a nil pointer is a programming error, and it not recoverable.